### PR TITLE
Make Iterator.from spec compliant

### PR DESCRIPTION
**Pull request by saghul**
Create a proper wrapper around the inner iterator and proxy the calls to next / return.

### CODE REVIEW

File: test262_errors.txt [hunk 1]
```diff
@@ -42,14 +42,6 @@ test262/test/built-ins/Date/prototype/setUTCMonth/date-value-read-before-tonumbe
 test262/test/built-ins/Date/prototype/setUTCMonth/date-value-read-before-tonumber-when-date-is-invalid.js:26: strict mode: Test262Error: time updated in valueOf Expected SameValue(«NaN», «0») to be true
 test262/test/built-ins/Date/prototype/setUTCSeconds/date-value-read-before-tonumber-when-date-is-invalid.js:26: Test262Error: time updated in valueOf Expected SameValue(«NaN», «0») to be true
 test262/test/built-ins/Date/prototype/setUTCSeconds/date-value-read-before-tonumber-when-date-is-invalid.js:26: strict mode: Test262Error: time updated in valueOf Expected SameValue(«NaN», «0») to be true
-test262/test/built-ins/Iterator/from/get-return-method-when-call-return.js:24: TypeError: not a function
```
Comment (saghul) [inline test262_errors.txt:45]: These were the new ones after the test262 update. There are a couple of more failures below in weird property access, but as of this PR Iterator.from is non-ghetto :-P

File: quickjs.c [hunk 2]
```diff
@@ -40053,24 +40057,95 @@ static JSValue js_array_iterator_next(JSContext *ctx, JSValue this_val,
     }
 }
 
+typedef struct JSIteratorData {
```
Comment (saghul) [inline quickjs.c]: @bnoordhuis Let me know what you think of this approach, I couldn't find a way around not having extra stored state.

**Comment by saghul** [comment]
Doh, I'm an idiot. Looks like we do need a new class because after fixing the prototype whoes I cannot make test262/test/built-ins/Iterator/from/result-proto.js pass.

New class it is!

**Comment by saghul** [comment]
Updated, PTAL! Note it includes https://github.com/quickjs-ng/quickjs/pull/690

File: quickjs-atom.h [hunk 3]
```diff
@@ -218,6 +218,7 @@ DEF(WeakMap, "WeakMap") /* Map + 2 */
 DEF(WeakSet, "WeakSet") /* Map + 3 */
 DEF(Iterator, "Iterator")
 DEF(IteratorHelper, "Iterator Helper")
+DEF(IteratorWrap, "Iterator Wrap")
```
Comment (bnoordhuis) [inline quickjs-atom.h:221]: I did wonder in that other PR why there was no Wrap class.

File: quickjs.c [hunk 4]
```diff
@@ -40103,29 +40163,29 @@ static JSValue js_iterator_from(JSContext *ctx, JSValue this_val,
...
-        // define a separate JS_CLASS_NON_GHETTO_ITERATOR
-        temp = method;
-        method = js_function_bind(ctx, method, 1, &obj);
-        JS_FreeValue(ctx, temp);
-        if (JS_IsException(method))
-            return JS_EXCEPTION;
-        iter = JS_NewObjectProtoClass(ctx, ctx->iterator_proto, JS_CLASS_ITERATOR);
+        iter = JS_NewObjectClass(ctx, JS_CLASS_ITERATOR_WRAP);
         if (JS_IsException(iter)) {
             JS_FreeValue(ctx, method);
```
Comment (bnoordhuis) [inline quickjs.c:40168]: Move this to `fail:` so you can just `goto fail;` here and below?

**Review by bnoordhuis — APPROVED** [review]

**Pull request merged**